File: typescript_challenges/01_ttlCache.ts
/**
 * Challenge 01 — Generic TTL Cache
 * ----------------------------------
 * Implement TTLCache<K, V>, a generic cache where every entry expires after
 * a fixed number of milliseconds.
 *
 * API:
 *   constructor(ttlMs: number)
 *   set(key: K, value: V): void       — stores value; resets TTL if key exists
 *   get(key: K): V | undefined        — undefined if absent or expired (lazy eviction)
 *   delete(key: K): boolean           — true if key existed and was removed
 *   size(): number                    — count of non-expired entries
 *   clear(): void
 *
 * Rules:
 *   - Lazy eviction: an expired entry is removed when get() is called for it.
 *   - size() must not count expired entries.
 *   - Thread safety is not required (single-threaded JS runtime).
 *
 * Usage:
 *   const cache = new TTLCache<string, number>(5_000);
 *   cache.set("hits", 42);
 *   cache.get("hits"); // 42
 *   // (after 5 seconds)
 *   cache.get("hits"); // undefined
 */
export class TTLCache<K, V> {
  constructor(private readonly ttlMs: number) {
    throw new Error("not implemented");
  }

  set(key: K, value: V): void {
    throw new Error("not implemented");
  }

  get(key: K): V | undefined {
    throw new Error("not implemented");
  }

  delete(key: K): boolean {
    throw new Error("not implemented");
  }

  size(): number {
    throw new Error("not implemented");
  }

  clear(): void {
    throw new Error("not implemented");
  }
}

File: go_challenges/01_ttl_cache/01_ttl_cache.go
// Challenge 01 — Generic TTL Cache
// ----------------------------------
// Implement TTLCache[K comparable, V any], a generic cache where each entry
// expires after a fixed duration.
//
// API:
//
//	NewTTLCache[K, V](ttl time.Duration) *TTLCache[K, V]
//
//	(c *TTLCache[K, V]) Set(key K, value V)
//	    Stores value under key. Resets the TTL if the key already exists.
//
//	(c *TTLCache[K, V]) Get(key K) (V, bool)
//	    Returns (value, true) if the key exists and has not expired.
//	    Returns (zero, false) otherwise. Expired entries are removed lazily.
//
//	(c *TTLCache[K, V]) Delete(key K)
//
//	(c *TTLCache[K, V]) Len() int
//	    Number of non-expired entries.
//
// Thread safety: Set, Get, Delete, and Len must be safe for concurrent use.
//
// Usage:
//
//	cache := NewTTLCache[string, int](5 * time.Second)
//	cache.Set("hits", 42)
//	v, ok := cache.Get("hits") // v=42, ok=true

package main

import (
	"sync"
	"time"
)

type TTLCache[K comparable, V any] struct {
	// your fields here
}

func NewTTLCache[K comparable, V any](ttl time.Duration) *TTLCache[K, V] {
	panic("not implemented")
}

func (c *TTLCache[K, V]) Set(key K, value V) {
	panic("not implemented")
}

func (c *TTLCache[K, V]) Get(key K) (V, bool) {
	panic("not implemented")
}

func (c *TTLCache[K, V]) Delete(key K) {
	panic("not implemented")
}

func (c *TTLCache[K, V]) Len() int {
	panic("not implemented")
}

// keep imports used
var (
	_ sync.RWMutex
	_ = time.Second
)

func main() {
	cache := NewTTLCache[string, int](2 * time.Second)
	cache.Set("a", 1)
	v, ok := cache.Get("a")
	fmt.Printf("Get a: %v, found: %v\n", v, ok)
	time.Sleep(3 * time.Second)
	v, ok = cache.Get("a")
	fmt.Printf("Get a after TTL: %v, found: %v\n", v, ok)
}
